### PR TITLE
Spatial System rewrite

### LINKED ISSUE #446: [BUG] Manually exporting a prefab crashes in the spatial system
Open the "Box" prefab, edit anything (mesh color or such). Then CTRL+E and "Export only" -> crash in 
`void UpdateObject(ezSpatialData* pData)`

```
 	[Inline Frame] ezCore.dll!ezArrayBase<ezDynamicArray<ezSpatialSystem_RegularGrid::Cell::ObjectPointerAndFrame,ezDefaultAllocatorWrapper>,ezDynamicArrayBase<ezDynamicArray<ezSpatialSystem_RegularGrid::Cell::ObjectPointerAndFrame,ezDefaultAllocatorWrapper>>>::operator[](unsigned int uiIndex) Line 84	C++
>	[Inline Frame] ezCore.dll!ezSpatialSystem_RegularGrid::Cell::UpdateObject(ezSpatialData * pData) Line 285	C++
 	ezCore.dll!ezSpatialSystem_RegularGrid::SpatialDataObjectChanged(ezSpatialData * pData) Line 704	C++
 	ezCore.dll!ezSpatialSystem::UpdateSpatialData(const ezSpatialDataHandle & hData, const ezSimdBBoxSphere & bounds, ezGameObject * pObject, unsigned int uiCategoryBitmask) Line 107	C++
 	ezCore.dll!ezGameObject::operator=(const ezGameObject & other) Line 275	C++
 	[Inline Frame] ezCore.dll!ezMemoryUtils::Relocate(ezGameObject *) Line 457	C++
 	[Inline Frame] ezCore.dll!ezMemoryUtils::Relocate(ezGameObject *) Line 140	C++
 	[Inline Frame] ezCore.dll!ezBlockStorage<ezGameObject,16384,0>::Delete(ezGameObject *) Line 238	C++
 	ezCore.dll!ezBlockStorage<ezGameObject,16384,0>::Delete(ezGameObject * pObject, ezGameObject * & out_pMovedObject) Line 206	C++
 	ezCore.dll!ezWorld::DeleteDeadObjects() Line 1254	C++
 	ezCore.dll!ezWorld::Update() Line 540	C++
 	ezCore.dll!ezWorld::UpdateFromThread() Line 971	C++
 	ezFoundation.dll!ezTask::Run(unsigned int uiInvocation) Line 55	C++
 	ezFoundation.dll!ezTaskSystem::ExecuteTask(ezTaskPriority::Enum FirstPriority, ezTaskPriority::Enum LastPriority, bool bOnlyTasksThatNeverWait, const ezTaskGroupID & WaitingForGroup, ezAtomicInteger<int> * pWorkerState) Line 133	C++
 	ezFoundation.dll!ezTaskSystem::HelpExecutingTasks(const ezTaskGroupID & WaitingForGroup) Line 216	C++
 	ezFoundation.dll!ezTaskSystem::WaitForGroup(ezTaskGroupID Group) Line 285	C++
 	ezGameEngine.dll!ezGameApplication::UpdateWorldsAndExtractViews() Line 210	C++
 	ezFoundation.dll!ezTask::Run(unsigned int uiInvocation) Line 55	C++
 	ezFoundation.dll!ezTaskSystem::ExecuteTask(ezTaskPriority::Enum FirstPriority, ezTaskPriority::Enum LastPriority, bool bOnlyTasksThatNeverWait, const ezTaskGroupID & WaitingForGroup, ezAtomicInteger<int> * pWorkerState) Line 133	C++
 	ezFoundation.dll!ezTaskWorkerThread::Run() Line 68	C++
 	ezFoundation.dll!RunThread(ezThread * pThread) Line 55	C++
```

### LINKED ISSUE #454: [BUG] ezGameObject::GetNumFramesSinceVisible() doesn't work as expected with camera exclusion tags
Thought I add a ticket, because I had already forgotten about this issue again :)

ezGameObject::GetNumFramesSinceVisible() doesn't return a useful value, since the system ignores camera include/exclude tags and therefore things that actually don't get rendered, will still be marked as "recently visible".

### PR DESCRIPTION
**Pull request by C-Core**
Rewrote the Spatial System and the corresponding regular grid implementation to clean it up and add tag filtering directly to the spatial queries:
* The new system uses one grid per spatialdata category. This makes it much more efficient for queries with only a few objects in one category like e.g. wind volumes.
* To further improve performance the new system can create cached grids for frequent requested tag combinations. These cached grids then only contain the objects that match the queried tag combination. E.g. skylight visibility queries are speed up drastically since only very few objects in a scene contribute to skylight. Also shadow visibility queries can benefit from that if the scene contains many non shadow casting objects.

Fixes #446
Fixes #454

### CODE REVIEW

File: Code/Engine/Core/World/SpatialData.h [hunk 1]
```diff
@@ -1,11 +1,28 @@
...
+struct ezSpatialData
 {
+  struct Flags
+  {
+    typedef ezUInt8 StorageType;
+
+    enum Enum
+    {
+      None = 0,
+      FrequentChanges = EZ_BIT(0),
```
Comment (jankrassnigg) [inline Code/Engine/Core/World/SpatialData.h]: Comment what effect this has, please.

File: Code/Engine/Foundation/Math/Implementation/Math_inl.h [hunk 2]
```diff
@@ -80,6 +80,22 @@ namespace ezMath
 #endif
   }
 
+  EZ_ALWAYS_INLINE ezUInt32 FirstBitLow(ezUInt64 value)
```
Comment (jankrassnigg) [inline Code/Engine/Foundation/Math/Implementation/Math_inl.h:83]: Didn't i just add such a function in my arr merge?

File: Code/Engine/GameEngine/Effects/Wind/Implementation/WindVolumeComponent.cpp [hunk 3]
```diff
@@ -8,7 +8,7 @@
 #include <Foundation/Math/Intersection.h>
 #include <GameEngine/Effects/Wind/WindVolumeComponent.h>
 
-ezSpatialData::Category ezWindVolumeComponent::SpatialDataCategory = ezSpatialData::RegisterCategory("WindVolumes");
+ezSpatialData::Category ezWindVolumeComponent::SpatialDataCategory = ezSpatialData::RegisterCategory("WindVolumes", ezSpatialData::Flags::None);
```
Comment (jankrassnigg) [inline Code/Engine/GameEngine/Effects/Wind/Implementation/WindVolumeComponent.cpp:12]: Should they be separated by dynamic and static volumes?

**Review by jankrassnigg — APPROVED** [review]

File: Code/Engine/Foundation/Math/Implementation/Math_inl.h (hunk 2 above)
Comment (C-Core) [reply Code/Engine/Foundation/Math/Implementation/Math_inl.h:83]: Nope that was a different function

File: Code/Engine/GameEngine/Effects/Wind/Implementation/WindVolumeComponent.cpp (hunk 3 above)
Comment (C-Core) [reply Code/Engine/GameEngine/Effects/Wind/Implementation/WindVolumeComponent.cpp:12]: No, that only makes sense if you expect a lot of objects (like a few 1000 or so) in this category

**Pull request merged**